### DISCUSSION TITLE
Endpoints available for GitHub Apps / Accessing API endpoints as an installation

### DISCUSSION THREAD

**Discussion by joshkautz** [community/community#38215] [Apps, API and Webhooks] [answered]
I am trying to create a GitHub App that will create a new repository for users that install the GitHub App on their personal accounts. I have read the entirety of the [Building GitHub Apps](https://docs.github.com/en/developers/apps/building-github-apps) documentation several times over, and I am starting to arrive at the conclusion that it may not be possible for me to access the [**POST** /user/repos](https://docs.github.com/en/rest/repos/repos#create-a-repository-for-the-authenticated-user) API endpoint while authenticated as an installation on GitHub Apps, using an installation access token. Hopefully I'm using all the terminology properly here.

Ultimately, I am looking if anyone can confirm that it is or isn't possible for this behavior to be achieved. Is it possible for a GitHub App to create a new repository for a user that installs the GitHub App on their personal account? Is there an alternative way that I might consider to achieve this behavior? Thank you in advance; I'm happy to entertain any other solutions that folks might have.

_Edit:_
I am able to successfully generate a JWT using my GitHub App private key, and then use it in the authorization header in a request to the [**POST** /app/installations/\<Installation ID\>/access_tokens](https://docs.github.com/en/rest/apps/apps#create-an-installation-access-token-for-an-app) endpoint to generate an access token on behalf of an installation of the GitHub App.

**Request**
```
curl -i -X POST -H "Authorization: Bearer <JWT>" -H "Accept: application/vnd.github+json" https://api.github.com/app/installations/<Installation ID>/access_tokens
```
**Response**
```json
{
  "token": "<Access Token>",
  "expires_at": "2022-11-03T21:56:47Z",
  "permissions": {
    "administration": "write",
    "metadata": "read"
  },
  "repository_selection": "all"
}
```

When I try using this access token to create a new repository on the installation, with the [**POST** /user/repos](https://docs.github.com/en/rest/repos/repos#create-a-repository-for-the-authenticated-user) endpoint, I receive a response with a 403 status code and error message, included below:

```
curl -i -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Token <Access Token>" https://api.github.com/user/repos -d '{"name":"Hello-World"}'
```

```json
{
  "message": "Resource not accessible by integration",
  "documentation_url": "https://docs.github.com/rest/reference/repos#create-a-repository-for-the-authenticated-user"
}
```

Hopefully this is helpful for anyone kind enough to take the time to help me out with this. Once again, tremendous thank you in advance!

**Comment by steffen** [comment, marked as answer]
👋 @joshkautz That endpoint does not work with GitHub Apps. Look out for the `Works with GitHub Apps` label in the documentation that is shown on each endpoint where GitHub Apps are supported.

That [endpoint's documentation](https://docs.github.com/en/rest/repos/repos#create-a-repository-for-the-authenticated-user) also says "for the authenticated user", but you are authenticated with an GitHub Apps installation token.

You can create repositories for an organization though (on which you have installed the GitHub App):
https://docs.github.com/en/rest/repos/repos#create-an-organization-repository

Also for the record, here is a complete list of supported endpoints for GitHub Apps:
https://docs.github.com/en/rest/overview/endpoints-available-for-github-apps

Honestly, I'm a little bit surprised that there is no REST API endpoint for creating user repositories for GitHub Apps. I assume an endpoint would need to look like `POST /users/{username}/repos` similar to this one:
https://docs.github.com/en/rest/repos/repos#list-repositories-for-a-user

Having said that, but it's worth giving the GraphQL API and its [`createRepository`](https://docs.github.com/en/graphql/reference/mutations#createrepository) mutation a try?

I hope this helps!

**Reply by joshkautz** [reply]
Thanks, @steffen ! Even if it's not what I wanted to hear, it is indeed helpful to have someone confirm my suspicions. Hopefully in the future a REST API endpoint for creating user repositories for GitHub Apps will become available. I'll take a look at the GraphQL API in the meantime. Have a good one!

**Reply by joshkautz** [reply]
I ended up creating an OAuth App, and going through the authorization flow to get the OAuth Code, use it to generate an Access Token and authenticate on behalf of the user, and using that token to successfully use the REST API endpoint for creating a new repository. This can be done with a GitHub app also when using OAuth Authorization flow, but I liked the flow of using an OAuth App instead.
